bftengine/src/bftengine/FullNodeReplica.cpp

```diff
@@ -0,0 +1,295 @@
+  }
+  fn_metrics_.received_checkpoint_msg_++;
+  LOG_INFO(GL,
+           KVLOG(msg->senderId(),
+                 msg->idOfGeneratedReplica(),
+                 msg->seqNumber(),
+                 msg->epochNumber(),
+                 msg->size(),
+                 msg->isStableState(),
+                 msg->state(),
+                 msg->stateDigest(),
+                 msg->reservedPagesDigest(),
+                 msg->rvbDataDigest()));
+
+  // Reconfiguration cmd block is synced to RO replica via reserved pages
+  EpochNum replicasLastKnownEpochVal = 0;
+  auto epochNumberFromResPages = ReconfigurationCmd::instance().getReconfigurationCommandEpochNumber();
+  if (epochNumberFromResPages.has_value()) replicasLastKnownEpochVal = epochNumberFromResPages.value();
+
+  // not relevant
+  if (!msg->isStableState() || msg->seqNumber() <= lastExecutedSeqNum ||
+      msg->epochNumber() < replicasLastKnownEpochVal) {
+    return;
+  }
+  // no self certificate
+  static std::map<SeqNum, CheckpointInfo<false>> checkpointsInfo;
+  const auto msgSeqNum = msg->seqNumber();
+  const auto idOfGeneratedReplica = msg->idOfGeneratedReplica();
+  checkpointsInfo[msgSeqNum].addCheckpointMsg(msg.release(), idOfGeneratedReplica);
+  // if enough - invoke state transfer
+  if (checkpointsInfo[msgSeqNum].isCheckpointCertificateComplete()) {
+    persistCheckpointDescriptor(msgSeqNum, checkpointsInfo[msgSeqNum]);
+    checkpointsInfo.clear();
+    LOG_INFO(GL, "call to startCollectingState()");
+    stateTransfer->startCollectingState();
+  }
+}
+
+void FullNodeReplica::persistCheckpointDescriptor(const SeqNum &seqnum, const CheckpointInfo<false> &chckpinfo) {
+  std::vector<CheckpointMsg *> msgs;
+  msgs.reserve(chckpinfo.getAllCheckpointMsgs().size());
+  for (const auto &m : chckpinfo.getAllCheckpointMsgs()) {
+    msgs.push_back(m.second);
+    LOG_INFO(GL,
+             KVLOG(m.second->seqNumber(),
+                   m.second->epochNumber(),
+                   m.second->state(),
+                   m.second->stateDigest(),
+                   m.second->reservedPagesDigest(),
+                   m.second->rvbDataDigest(),
+                   m.second->idOfGeneratedReplica()));
+  }
+  DescriptorOfLastStableCheckpoint desc(ReplicaConfig::instance().getnumReplicas(), msgs);
+  const size_t bufLen = DescriptorOfLastStableCheckpoint::maxSize(ReplicaConfig::instance().getnumReplicas());
+  concord::serialize::UniquePtrToChar descBuf(new char[bufLen]);
+  char *descBufPtr = descBuf.get();
+  size_t actualSize = 0;
+  desc.serialize(descBufPtr, bufLen, actualSize);
+  ConcordAssertNE(actualSize, 0);
+
+  // TODO [TK] S3KeyGenerator
+  // checkpoints/<BlockId>/<RepId>
+  std::ostringstream oss;
+  oss << "checkpoints/" << msgs[0]->state() << "/" << config_.replicaId;
+  metadataStorage_->atomicWriteArbitraryObject(oss.str(), descBuf.get(), actualSize);
+}
+
+template <>
+void FullNodeReplica::onMessage<ClientRequestMsg>(std::unique_ptr<ClientRequestMsg> msg) {
```

Comment:
This comment is really important to my idea: 
How much of the code is fully duplicated?
If there is a significant amount of code duplicated, why not create a base class and bring this common code up, to reuse it on both RO/FN replicas?
For this function, it looks identical in logic, Are there other functions and members where you only changed comments or logs and the code is the same?

Reply:
Hi @glevkovich , FullNodeReplica.cpp and  FullNodeReplica.hpp are exact copy of ReadOnlyReplica.cpp and ReadOnlyReplica.hpp with class name change

Reply:
other than the above files all other files are either modified or added